metacheck/main.c

```diff
@@ -0,0 +1,406 @@
+	msg.o.size = size;
+
+	idevctl->type = flashsrv_devctl_readraw;
+	idevctl->readraw.oid = oid;
+	idevctl->readraw.size = size;
+	idevctl->readraw.address = addr;
+
+	if (msgSend(oid.port, &msg) < 0) {
+		return -1;
+	}
+
+	if (odevctl->err != size) {
+		return odevctl->err;
+	}
+
+	return 0;
+}
+
+
+static int containsCleanmarker(const uint8_t *buf)
+{
+	static const uint8_t cleanmarker[] = { 133, 25, 3, 32, 8, 0, 0, 0, 255, 255, 255, 255, 255, 255, 255, 255 };
+
+	return !memcmp(buf, cleanmarker, sizeof(cleanmarker));
+}
+
+
+static int isBadBlock(const uint8_t *buf)
+{
+	if (buf[0] == 0 && buf[1] == 0) {
+		return 1;
+	}
+
+	return 0;
+}
+
+
+static int isErased(const uint8_t *buf)
+{
+	int i;
+
+	for (i = 0; i < META_SZ; i++) {
+		if (buf[i] != 0xFF) {
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+
+static int checkErasedECC(const uint8_t *buf)
+{
+	int i;
+
+	buf += META_SZ;
+	for (i = 0; i < METAECC_SZ; i++) {
+		if (buf[i] != 0xff) {
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+
+static uint8_t reverse_bit(uint8_t in_byte)
+{
+	int i;
+	uint8_t out_byte = 0;
+
+	for (i = 0; i < 8; i++) {
+		if (in_byte & ((0x80) >> i)) {
+			out_byte |= 1 << i;
+		}
+	}
+
+	return out_byte;
+}
+
+
+static void encodeMeta(struct bch_control *bch, const uint8_t *inbuf, uint8_t *outbuf)
+{
+	int i;
+
+	memset(outbuf, 0, METAECC_SZ);
+	encode_bch(bch, inbuf, META_SZ, outbuf);
+
+	for (i = 0; i < METAECC_SZ; i++) {
+		outbuf[i] = reverse_bit(outbuf[i]);
+	}
+}
+
+
+static void dumpECC(partition_t *part, const char *type, const uint8_t *inbuf, size_t offset)
+{
+	size_t block = offset / part->rawblocksz;
+	size_t page = (offset % part->rawblocksz) / part->rawpagesz;
+	int i;
+
+	printf("Bad meta ECC (marker: %6s, block: %3u, page: %2u): ", type, block, page);
+	inbuf += META_SZ;
+	for (i = 0; i < METAECC_SZ; i++) {
+		printf("%02x ", inbuf[i]);
+	}
+	putchar('\n');
+}
+
+
+static void dumpWeirdMeta(partition_t *part, size_t offset, const uint8_t *inbuf)
+{
+	size_t block = offset / part->rawblocksz;
+	size_t page = (offset % part->rawblocksz) / part->rawpagesz;
+	int i;
+
+	printf("Weird metadata (block: %u: page: %u): ", offset / part->rawblocksz,
+		(offset % part->rawblocksz) / part->rawpagesz);
+
+	for (i = 0; i < META_SZ; i++) {
+		printf("%02x ", inbuf[i]);
+	}
+	putchar('\n');
+}
+
+
+static int checkPartition(partition_t *part, struct bch_control *bch)
+{
+	uint8_t eccbuf[METAECC_SZ];
+	int totalcnt = 0, goodcnt = 0, badecc = 0, badcnt = 0;
+	int erasedcnt = 0, weirdcnt = 0;
+	int ret, eccInvalid;
+	size_t offset = 0;
+	int i;
+
+	ret = readraw(part->oid, 0, common.buf, part->rawpagesz);
+	while (ret == 0) {
+		encodeMeta(bch, common.buf, eccbuf);
+		eccInvalid = memcmp(eccbuf, common.buf + META_SZ, METAECC_SZ);
+
+		if (containsCleanmarker(common.buf)) {
+			if (eccInvalid) {
+				badecc++;
+				if (common.verbose) {
+					dumpECC(part, "clean", common.buf, offset);
+				}
+			}
+			else {
+				goodcnt++;
+			}
+		}
+		else if (isBadBlock(common.buf)) {
```

Comment:
bad blocks should be less frequent then erased ones, I think, so maybe swap these cases around?

Reply:
That's right, thanks.